Configure new agent for projects build

AAndrisa opened the pull request.

amiclaus reviewed (commented):
Is this PR related to https://github.com/analogdevicesinc/no-OS/pull/1746 ?
If so, please provide some info / reference to it and what was added on top.
Can we close the old one?

SRaus commented on tools/scripts/download_files.py:
```diff
@@ -0,0 +1,56 @@
+
+NOOS_PATH = sys.argv[1]
+BUILD_PATH = sys.argv[2]
+SERVER_BASE_PATH = sys.argv[3]
```
SERVER_BASE_PATH is actually the SERVER BASE PATH plus the HDL folder. 
You can rename it accordingly and pay attention on the value given in azure - it should contain also 'hdl' folder (and most probably the branch name).

AAndrisa commented:
This PR is related to #1746, it was forked from that branch and added on top the script for downloading the hardware files from artifactory and some small changes related to error status for stm32 projects and eliminating the builds.json files of the projects with no hardware.

SRaus commented:
> was forked from that branch
Ok. so we can close the other PR and focus just on this to get merged. 
PS.: The fork is done from other repo, not from a branch. But I've got the idea.

buha commented on tools/scripts/build_projects.py:
```diff
@@ -239,7 +239,7 @@ def build(self):
 			if err != 0:
 				return err
-			err = run_cmd_stm("timeout 200s " + cmd + ' VERBOSE=y -j%d all' % (multiprocessing.cpu_count() - 1))
+			err = run_cmd_stm("timeout 200s " + cmd + ' VERBOSE=y -j%d all' % ((multiprocessing.cpu_count() - 1)/2))
```
let's skip the `- 1` then, just divide by 2 if you think "all but one" cores is too much

amiclaus reviewed (commented):
for commit f62f3dce32b0e4ac1b2cec23f9b810452730bb03 even though it is not directly yours, please add a commit description explaining what happens. This is quite a big change in the build infrastructure.

amiclaus reviewed (commented):
The artifacts are now split per each job (platform).
![image](https://github.com/analogdevicesinc/no-OS/assets/26061529/537ea392-e667-4b26-b718-6295df6f0c48)
We need to somehow differentiate between platforms (maybe add suggestive name inside each string). otherwise it will be pretty unpleasant for the user to obtain a specific log.

amiclaus commented on tools/scripts/build_projects.py:
```diff
@@ -112,7 +112,7 @@ def run_cmd_stm(cmd):
 		ERR = 1
 		return err
 	err = os.system(cmd + ' >> %s 2>&1' % log_file)
-	if err != 0 and err != 31744 and tmp[0] != 'cp':
```
this looks like a revert of https://github.com/analogdevicesinc/no-OS/commit/4048b1ea4249104937c319b5efaefb86a8645cb5 ? Now there is another error treated (`124`). Maybe add separate commit explaining why and what you did. It is important to keep good track of these infrastructure changes.

amiclaus reviewed (commented):
there is a empty file added `log.txt`. please remove it.

amiclaus commented on tools/scripts/build_projects.py:
```diff
@@ -132,22 +132,6 @@ def to_blue(str):
 NEW_HW_DIR_NAME = 'new_hardware'
-FILE_TO_DOWNLOAD = 'latest.zip'
-
-def download_all_hw(builds_dir):
```
we should keep things in modules. Looks like you moved the new download flow inside main function. I'd suggest keeping it in a separate function as it was until now.

amiclaus reviewed (commented):
you changed the cpu_count handling and error handling. These changes look like optimization rather then "new agent configuration". I would do separate commit for cpu_count updates (explaining why in the commit description) and separate commit for error handling (also explaining why in the commit description).

amiclaus commented on tools/scripts/download_files.py:
```diff
@@ -0,0 +1,61 @@
+                    os.system("mkdir -p %s" % (str(new_harware_dir) + '/' + hardware))
+                    os.system("wget -q -nv -P %s %s" % (str(new_harware_dir) + '/' + hardware, d_path))
+                    #if you have an error regarding properties is because of python version, it works just with python3.8 or older
+                    #but if you want to use the script just delete the next 3 lines and remove the variabile from the printed message
```
we should not encourage users to delete lines of code from the existing scripts to make things work. I'd rather parameterize the script. Or check inside this script the python version used by the host and throw a warning if the versions are not compatible.

AAndrisa commented:
Changes added: added different commits for two new topics (cpu_count and status of timeout) and also changed the handling of the message in case of missing project from latest timestamp, and also changed the way donwload_files.py script is called.

amiclaus reviewed (commented):
looks good! I re-triggered the Xilinx build since it appeared as canceled. Also, I think we need to sync the merge of this PR with the project build fixes related to the new .xsa downloads.

amiclaus reviewed (commented):
performed a rebase with latest master. The ADucM build issues should be fixed now. No changes were made in the current commits of this PR.

amiclaus reviewed (approved):
with the early mention, that xilinx build should be fixed.
